pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -446,52 +454,98 @@ private String getOriginalPrincipal(String originalAuthData, String originalAuth
                     // Legacy client is passing enum
                     authMethod = connect.getAuthMethod().name().substring(10).toLowerCase();
+                } else {
+                    authMethod = "none";
+                }
+
+                authenticationProvider = getBrokerService()
+                    .getAuthenticationService()
+                    .getAuthenticationProvider(authMethod);
+
+                // Not find provider named authMethod. Most used for tests.
+                // In AuthenticationDisabled, it will set authMethod "none".
+                if (authenticationProvider == null) {
+                    authRole = getBrokerService().getAuthenticationService().getAnonymousUserRole();
+                    completeConnect(connect);
+                    return;
                 }
 
-                String authData = connect.getAuthData().toStringUtf8();
+                // init authState and other var
                 ChannelHandler sslHandler = ctx.channel().pipeline().get(PulsarChannelInitializer.TLS_HANDLER);
                 SSLSession sslSession = null;
                 if (sslHandler != null) {
                     sslSession = ((SslHandler) sslHandler).engine().getSession();
                 }
                 originalPrincipal = getOriginalPrincipal(
-                        connect.hasOriginalAuthData() ? connect.getOriginalAuthData() : null,
-                        connect.hasOriginalAuthMethod() ? connect.getOriginalAuthMethod() : null,
-                        connect.hasOriginalPrincipal() ? connect.getOriginalPrincipal() : null,
-                        sslSession);
-                authenticationData = new AuthenticationDataCommand(authData, remoteAddress, sslSession);
-                authRole = getBrokerService().getAuthenticationService()
-                        .authenticate(authenticationData, authMethod);
-
-                log.info("[{}] Client successfully authenticated with {} role {} and originalPrincipal {}", remoteAddress, authMethod, authRole, originalPrincipal);
-            } catch (AuthenticationException e) {
-                String msg = "Unable to authenticate";
-                log.warn("[{}] {}: {}", remoteAddress, msg, e.getMessage());
-                ctx.writeAndFlush(Commands.newError(-1, ServerError.AuthenticationError, msg));
-                close();
+                    connect.hasOriginalAuthData() ? connect.getOriginalAuthData() : null,
+                    connect.hasOriginalAuthMethod() ? connect.getOriginalAuthMethod() : null,
+                    connect.hasOriginalPrincipal() ? connect.getOriginalPrincipal() : null,
+                    sslSession);
+
+                authenticationData = authenticationProvider.getAuthDataSource(clientData, remoteAddress, sslSession);
+                authState = authenticationProvider.newAuthState(authenticationData);
+            }
+
+            AuthData brokerData = authState.authenticate(clientData);
```

Comment:
Having this return null is a little weird, as null is a sort of negative signal. I think it would be better to have a method on AuthData, isAuthenticationComplete(), which then can be used to decide which message to send.

Reply:
Thanks, @ivankelly ， isComplete() is more like a method in authState, will add it in authState.